pkg/csi/service/vanilla/controller.go

```diff
@@ -596,77 +597,151 @@ func (c *controller) createBlockVolume(ctx context.Context, req *csi.CreateVolum
+			log.Infof("Volume with name %q and id %q is already created on CNS with opId: %q.",
+				req.Name, volumeOperationDetails.VolumeID, volumeOperationDetails.OperationDetails.OpID)
+
+			volumeInfo = &cnsvolume.CnsVolumeInfo{
+				DatastoreURL: "",
+				VolumeID: cnstypes.CnsVolumeId{
+					Id: volumeOperationDetails.VolumeID,
+				},
+			}
+			volTaskAlreadyRegistered = true
+		} else if volumeOperationDetails.OperationDetails.TaskStatus ==
+			cnsvolumeoperationrequest.TaskInvocationStatusInProgress &&
+			volumeOperationDetails.OperationDetails.TaskID != "" {
+			// If task is created in CNS for this volume but task is in progress, then
+			// we need to monitor the task to check if volume creation is completed or not.
+			log.Infof("Volume with name %s has CreateVolume task %s pending on CNS.",
+				req.Name, volumeOperationDetails.OperationDetails.TaskID)
+
+			taskMoRef := types.ManagedObjectReference{
+				Type:  "Task",
+				Value: volumeOperationDetails.OperationDetails.TaskID,
+			}
+			task := object.NewTask(vcenter.Client.Client, taskMoRef)
 
-		// Get shared accessible datastores for matching topology requirement.
-		if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.TopologyPreferentialDatastores) {
-			vcenter, err := c.manager.VcenterManager.GetVirtualCenter(ctx, c.manager.VcenterConfig.Host)
+			volumeInfo, faultType, err = c.manager.VolumeManager.MonitorCreateVolumeTask(ctx,
+				&volumeOperationDetails, task, req.Name, c.manager.CnsConfig.Global.ClusterID)
 			if err != nil {
-				return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
-					"failed to get vCenter. Err: %v", err)
-			}
-			sharedDatastores, err = c.topologyMgr.GetSharedDatastoresInTopology(ctx,
-				commoncotypes.VanillaTopologyFetchDSParams{
-					TopologyRequirement: topologyRequirement,
-					Vc:                  vcenter,
-					StoragePolicyName:   scParams.StoragePolicyName,
-				})
-			if err != nil || len(sharedDatastores) == 0 {
-				return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
-					"failed to get shared datastores for topology requirement: %+v. Error: %+v",
-					topologyRequirement, err)
+				return nil, faultType, logger.LogNewErrorCodef(log, codes.Internal,
+					"failed to monitor task for volume %s. Error: %+v", req.Name, err)
+			}
+			// Persist the operation details.
+			if volumeOperationDetails != nil && volumeOperationDetails.OperationDetails != nil &&
+				volumeOperationDetails.OperationDetails.TaskStatus !=
+					cnsvolumeoperationrequest.TaskInvocationStatusInProgress {
+				err := operationStore.StoreRequestDetails(ctx, volumeOperationDetails)
+				if err != nil {
+					log.Warnf("failed to store CreateVolume details with error: %v", err)
+				}
+			}
+
+			volTaskAlreadyRegistered = true
+		}
+	} else if err != nil {
+		log.Debugf("failed to get CreateVolume operation details for volume %s. Error: %+v",
```

Comment:
since this is going to occur anyways for the first time, try changing `failed` to something that doesn't sound alarming, may confuse the folks who are triaging

Reply:
done